keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1338 @@
+kind: ResourceQuota
+metadata:
+  name: quota
+spec:
+  hard:
+    limits.iops: "200000"    // max for all containers in the namespace
+    limits.throughput: 200
+    requests.iops: "2000"    // max per request 
+    requests.throughput: 20
+    pods: "10"    // max number of pods allowed in the namespace
+apiVersion: v1
+kind: LimitRange
+metadata:
+  name: sample
+spec:
+  limits:
+  - max:
+      iops: 1500
+      throughput: 10
+    min:
+      iops: 10
+      throughput: 1
+    default:
+      iops: 100
+      throughput: 5
+    defaultRequest:
+      iops: 100
+      throughput: 20
+    type: Container  
+```
+
+
+IO provisioning should have similar issues to resize (except that we have to solve reducing IO). Please see limit storage consumption. Quota validation should be part of [func ValidatePersistentVolumeClaimUpdate](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L2209) flow by using [func ValidateResourceQuota](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5758) and [func ValidateLimitRange](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5800).
+
+For ResourceQuota, we will verify that sum of spec.resources[iops] and spec.resources[throughput] for all PVCs in the Namespace from DSW don't exceed quota, for LimitRanger we check that a modify request does not violate the min and max limits specified in LimitRange for the pvc's namespace.
+
+
+##### Pros:
+
+*   Simplify user experience by giving only restricted, well defined controls
+
+##### Cons:
+
+
+*   Difficult to get consensus of what is iops/throughput among different storage providers
+*   Not all the storage providers support independently configurable iops/throughput
+
+
+#### Option 2: Opaque map in CreateVolume and ModifyVolume requests by end users
+
+The users will set the volume performance parameters directly in the PVC:
+
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  volumePerformanceParameters:
+    iops: "10000"
+    throughput: "100Mib/s"
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+
+The ModifyVolumeRequest in CSI spec will stay as:
+
+
+```
+message ModifyVolumeRequest {
+  // Contains identity information for the existing volume.
+  // This field is REQUIRED.
+  string volume_id = 1
+  // This field is OPTIONAL.This allows the CO to specify the
+  // opaque volume parameters. 
+  map<string, string> volume_parameters = 2;
+}
+```
+
+
+
+##### Pros:
+
+
+
+*   Flexible to fit into all the cloud providers
+*   More flexibility to end users and no cluster administrator needs to be involved(also a con)
+
+
+##### Cons:
+
+*   More unpredictable behaviors because it is an opaque map. Compared to the recommended approach that the cluster administrator actually has the control over the values.
```

Comment:
The biggest con IMO is it's not portable across providers.

Reply:
Added.